### PR TITLE
Usability improvements for Azure DevOps feedback

### PR DESCRIPTION
**Pull request by dennis-sig**
- Update the existing comment instead of posting a new comment after every commit. The feedback already works like that in GitHub, but not yet in Azure DevOps.
- Mark the comment as fixed/resolved/closed if you have achieved your Sigrid objective.

### CODE REVIEW

File: sigridci/sigridci/azure_pull_request_report.py [hunk 1]
```diff
@@ -31,14 +32,21 @@ def generate(self, analysisId, feedback, options):
...
+            status = Objective.determineStatus(feedback, options)
+
+            if existingId == None:
+                self.callAzure("POST", self.buildRequestBody(feedbackFile, status), None)
+                UploadLog.log("Published feedback to Azure DevOps")
+            else:
+                self.callAzure("PATCH", self.buildRequestBody(feedbackFile, status), existingId)
+                UploadLog.log("Updated existing feedback in Azure DevOps")
         except urllib.error.HTTPError as e:
             UploadLog.log(f"Warning: Azure DevOps API error: {e.code} / {e.fp.read()}")
```
Comment (patveck) [inline sigridci/sigridci/azure_pull_request_report.py:51]: How about logging even more:
```suggestion
            UploadLog.log(f"Warning: Azure DevOps API error calling {e.url}: {e.code} ({e.reason}) / {e.fp.read()}")
```

By including `{e.url}`, we know which endpoint is involved (and we learn whether we constructed the URL correctly). `{e.reason}` might give information not in the body (not sure).

File: sigridci/sigridci/azure_pull_request_report.py [hunk 2]
```diff
@@ -47,25 +55,46 @@ def isSupported(self, options):
...
         baseURL = os.environ["SYSTEM_TEAMFOUNDATIONCOLLECTIONURI"]
         project = os.environ["SYSTEM_TEAMPROJECTID"]
         repo = os.environ["BUILD_REPOSITORY_NAME"]
         pr = os.environ["SYSTEM_PULLREQUEST_PULLREQUESTID"]
         version = self.AZURE_API_VERSION
-        return f"{baseURL}{project}/_apis/git/repositories/{repo}/pullRequests/{pr}/threads?api-version={version}"
 
-    def buildRequestBody(self, feedbackFile):
+        if threadId:
+            return f"{baseURL}{project}/_apis/git/repositories/{repo}/pullRequests/{pr}/threads/{threadId}?api-version={version}"
```
Comment (patveck) [inline sigridci/sigridci/azure_pull_request_report.py:83]: What's the reason we're using different versions? I think that newer versions typically _extend_ existing endpoints, so we could as well choose the lowest version that has this endpoint (incl. the query parameters we need).

**Review by patveck — CHANGES_REQUESTED** [review]

File: sigridci/sigridci/azure_pull_request_report.py [hunk 3]
```diff
@@ -47,25 +55,46 @@ def isSupported(self, options):
...
             options.runMode == RunMode.FEEDBACK_ONLY
 
-    def buildURL(self):
+    def findExistingSigridCommentThreadId(self):
+        existingThreads = json.load(self.callAzure("GET", None, None))
+
+        for thread in existingThreads["value"]:
+            for comment in thread["comments"]:
+                if comment["content"].startswith("# Sigrid"):
+                    return thread["id"]
```
[the remark this answers is not in the document]
Comment (dennis-sig) [reply sigridci/sigridci/azure_pull_request_report.py:64]: The Azure DevOps comments are not editable if you use the out-of-the-box instructions from the documentation. Also note the sorting in Azure is latest-on-top, which is different from GitHub and GitLab that both list comments in chronological order (i.e. oldest-on-top). It's indeed a heuristic, but I believe the heuristic is reasonably robust.

File: docs/reference/technology-support.md [hunk 4]
```diff
@@ -419,30 +419,33 @@ In addition to supporting 300+ technologies, Sigrid also supports various open s
...
+Note this section only described *third party* security analysis analyzers. The results from these analyzers are then combined with the results from Sigrid's own security ruleset before being displayed.
+
+| Display name           | Technologies                      | Notes                      |
+|------------------------|-----------------------------------|----------------------------|
+| Astrée                 | C, C++                            | Requires *Sigrid Security for Embedded Systems* license. |
+| Bandit                 | Python                            |
+| Checkmarx              | (many)                            | Requires *Sigrid Security Checkmarx* license. |
+| ErrorProne.NET         | C#                                |
+| ESLint                 | JavaScript, TypeScript            |
+| Facebook Infer         | Java                              | Disabled by default, can be enabled in configuration. |
```
[the remark this answers is not in the document]
Comment (dennis-sig) [reply docs/reference/technology-support.md:431]: Of course I more or less sneaked this documentation update in, but I believe what I wrote is correct: Infer does not run by default, but you can enable it by adding it to the `enabled_analyzers` option in the scope file.

[the remark this answers is not in the document]
Comment (patveck) [reply docs/reference/technology-support.md:431]: I'm afraid that's not the case: I checked, and Infer itself is not present in the environment where we run our checkers.

File: sigridci/sigridci/azure_pull_request_report.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (patveck) [reply sigridci/sigridci/azure_pull_request_report.py:64]: Ah, clear! I think that's fine.

File: sigridci/sigridci/azure_pull_request_report.py [hunk 5]
```diff
@@ -47,25 +55,46 @@ def isSupported(self, options):
...
+
+        return None
+
+    def callAzure(self, method, body, threadId):
+        request = urllib.request.Request(self.buildURL(threadId), json.dumps(body).encode("utf-8"))
+        request.method = method
+        request.add_header("Authorization", f"Bearer {os.environ['SYSTEM_ACCESSTOKEN']}")
```
[the remark this answers is not in the document]
Comment (dennis-sig) [reply sigridci/sigridci/azure_pull_request_report.py:71]: We do, in `isSupported`. Basically, checking that environment is how Sigrid CI determines whether or not it's running inside of Azure DevOps. If it's false, this whole thing doesn't run.

File: sigridci/sigridci/azure_pull_request_report.py (hunk 2 above)
Comment (dennis-sig) [reply sigridci/sigridci/azure_pull_request_report.py:83]: Made it more clear by directly using the constant.

**Review by patveck — APPROVED** [review]

**Pull request merged**